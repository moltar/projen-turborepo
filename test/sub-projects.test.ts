import { synthProjectSnapshot, createSubProject, createProject } from './util'

describe('TurborepoProject', () => {
  it('should add workspaces', () => {
    expect.assertions(1)

    const project = createProject()

    const subProjectDir = 'packages/baz'

    createSubProject({
      parent: project,
      outdir: subProjectDir,
    })

    const synth = synthProjectSnapshot(project)

    expect(synth['package.json'].workspaces).toStrictEqual([subProjectDir])
  })

  it('should add TypeScript path mappings when turned on', () => {
    expect.assertions(2)

    const project = createProject({
      pathMapping: true,
    })

    const subProjectBarDir = 'packages/bar'
    const subProjectBar = createSubProject({
      parent: project,
      outdir: subProjectBarDir,
    })

    const subProjectBazDir = 'packages/baz'
    createSubProject({
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
    })

    const synth = synthProjectSnapshot(project)

    expect(synth['packages/baz/tsconfig.json'].compilerOptions.baseUrl).toBe('.')
    expect(synth['packages/baz/tsconfig.json'].compilerOptions.paths).toStrictEqual({
      [subProjectBar.package.packageName]: ['../bar/src'],
    })
  })

  it('should add VS Code settings for ESLint', () => {
    expect.assertions(1)

    const project = createProject()

    const subProjectDir = 'packages/baz'
    createSubProject({
      parent: project,
      outdir: subProjectDir,
    })

    const synth = synthProjectSnapshot(project)

    expect(synth['.vscode/settings.json'].eslint).toBeDefined()
  })


  it('should set composite flag on the tsconfig', () => {
    expect.assertions(2)

    const project = createProject({ projectReferences: true })

    const subProjectBarDir = 'packages/bar'
    const subProjectBar = createSubProject({
      parent: project,
      outdir: subProjectBarDir,
    })

    const subProjectBazDir = 'packages/baz'
    createSubProject({
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
    })

    const synth = synthProjectSnapshot(project)

    expect(synth['tsconfig.json'].compilerOptions.composite).toBe(true)
    expect(synth['packages/bar/tsconfig.json'].compilerOptions.composite).toBe(true)
  })

  it('should add TypeScript project references when turned on', () => {
    expect.assertions(1)

    const project = createProject({
      projectReferences: true,
    })

    const subProjectBarDir = 'packages/bar'
    const subProjectBar = createSubProject({
      parent: project,
      outdir: subProjectBarDir,
    })

    const subProjectBazDir = 'packages/baz'
    createSubProject({
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
    })

    const synth = synthProjectSnapshot(project)

    expect(synth['packages/baz/tsconfig.json'].references).toStrictEqual([
      {
        path: '../bar',
      },
    ])
  })

  it('should add moduleNameMapper for jest', () => {
    expect.assertions(1)

    const project = createProject({
      jestModuleNameMapper: true,
    })

    const subProjectBarDir = 'packages/bar'
    const subProjectBar = createSubProject({
      parent: project,
      outdir: subProjectBarDir,
    })

    const subProjectBazDir = 'packages/baz'
    createSubProject({
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
    })

    const synth = synthProjectSnapshot(project)

    expect(synth['packages/baz/package.json'].jest.moduleNameMapper).toStrictEqual({
      [`^${subProjectBar.name}$`]: '<rootDir>/../bar/src',
    })
  })

  it('should have no build job in root for when parallelWorkflows are on', () => {
    expect.assertions(1)

    const project = createProject({
      parallelWorkflows: true,
      mutableBuild: false,
    })

    const subProjectBarDir = 'packages/bar'
    const subProjectBar = createSubProject({
      name: 'bar',
      parent: project,
      outdir: subProjectBarDir,
    })

    const subProjectBazDir = 'packages/baz'
    createSubProject({
      name: 'baz',
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
    })

    const synth = synthProjectSnapshot(project)

    expect(synth['.github/workflows/build.yml']).toMatchSnapshot()
  })

  it('should have one build job in root for when parallelWorkflows are off', () => {
    expect.assertions(1)

    const project = createProject({
      parallelWorkflows: false,
      mutableBuild: false,
    })
    const synth = synthProjectSnapshot(project)

    expect(synth['.github/workflows/build.yml']).toMatchSnapshot()
  })
})