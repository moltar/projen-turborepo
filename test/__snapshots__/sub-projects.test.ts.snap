// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TurborepoProject should have no build job in root for when parallelWorkflows are on 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: build
on:
  pull_request: {}
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    outputs:
      self_mutation_happened: \${{ steps.self_mutation.outputs.self_mutation_happened }}
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup Node.js
        uses: actions/setup-node@v2
        with:
          cache: yarn
          cache-dependency-path: \\"**/yarn.lock\\"
      - name: Install dependencies
        run: yarn install --check-files
      - name: build
        run: npx projen build
      - id: self_mutation
        name: Find mutations
        run: |-
          git add .
          git diff --staged --patch --exit-code > .repo.patch || echo \\"::set-output name=self_mutation_happened::true\\"
      - if: steps.self_mutation.outputs.self_mutation_happened
        name: Upload patch
        uses: actions/upload-artifact@v2
        with:
          name: .repo.patch
          path: .repo.patch
      - name: Fail build on mutation
        if: steps.self_mutation.outputs.self_mutation_happened
        run: |-
          echo \\"::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch.\\"
          cat .repo.patch
          exit 1
      - name: Upload artifact
        uses: actions/upload-artifact@v2.1.1
        with:
          name: build-artifact
          path: dist
  turbo:
    name: build
    needs: build
    runs-on: ubuntu-latest
    permissions:
      actions: read
      contents: read
    if: \\"! needs.build.outputs.self_mutation_happened\\"
    steps:
      - name: Download build artifacts
        uses: actions/download-artifact@v2
        with:
          name: build-artifact
          path: dist
      - name: Checkout
        uses: actions/checkout@v2
      - name: Setup turbo cache
        uses: felixmosh/turborepo-gh-artifacts@v1
        with:
          repo-token: \${{ secrets.GITHUB_TOKEN }}
          server-token: \${{ github.sha }}
      - name: Setup Node.js
        uses: actions/setup-node@v2
        with:
          cache: yarn
          cache-dependency-path: \\"**/yarn.lock\\"
      - name: Install dependencies
        run: yarn install --check-files --frozen-lockfile
      - name: Build
        run: npx turbo run build --api=\\"http://127.0.0.1:9080\\" --token=\\"\${{ github.sha }}\\" --team=\\"\${{ github.repository_owner }}\\" --scope=\\"\${{ matrix.scope }}\\" --include-dependencies
    strategy:
      fail-fast: true
      matrix:
        scope:
          - bar
          - baz
"
`;

exports[`TurborepoProject should have one build job in root for when parallelWorkflows are off 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: build
on:
  pull_request: {}
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    outputs:
      self_mutation_happened: \${{ steps.self_mutation.outputs.self_mutation_happened }}
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup Node.js
        uses: actions/setup-node@v2
        with:
          cache: yarn
          cache-dependency-path: \\"**/yarn.lock\\"
      - name: Setup turbo cache
        uses: felixmosh/turborepo-gh-artifacts@v1
        with:
          repo-token: \${{ secrets.GITHUB_TOKEN }}
          server-token: \${{ github.sha }}
      - name: Install dependencies
        run: yarn install --check-files
      - name: build
        run: npx projen build
      - id: self_mutation
        name: Find mutations
        run: |-
          git add .
          git diff --staged --patch --exit-code > .repo.patch || echo \\"::set-output name=self_mutation_happened::true\\"
      - if: steps.self_mutation.outputs.self_mutation_happened
        name: Upload patch
        uses: actions/upload-artifact@v2
        with:
          name: .repo.patch
          path: .repo.patch
      - name: Fail build on mutation
        if: steps.self_mutation.outputs.self_mutation_happened
        run: |-
          echo \\"::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch.\\"
          cat .repo.patch
          exit 1
      - name: Upload artifact
        uses: actions/upload-artifact@v2.1.1
        with:
          name: build-artifact
          path: dist
  turbo:
    name: build
    needs: build
    runs-on: ubuntu-latest
    permissions:
      actions: read
      contents: read
    if: \\"! needs.build.outputs.self_mutation_happened\\"
    steps:
      - name: Download build artifacts
        uses: actions/download-artifact@v2
        with:
          name: build-artifact
          path: dist
      - name: Checkout
        uses: actions/checkout@v2
      - name: Setup turbo cache
        uses: felixmosh/turborepo-gh-artifacts@v1
        with:
          repo-token: \${{ secrets.GITHUB_TOKEN }}
          server-token: \${{ github.sha }}
      - name: Setup Node.js
        uses: actions/setup-node@v2
        with:
          cache: yarn
          cache-dependency-path: \\"**/yarn.lock\\"
      - name: Setup turbo cache
        uses: felixmosh/turborepo-gh-artifacts@v1
        with:
          repo-token: \${{ secrets.GITHUB_TOKEN }}
          server-token: \${{ github.sha }}
      - name: Install dependencies
        run: yarn install --check-files --frozen-lockfile
      - name: Build
        run: npx turbo run build --api=\\"http://127.0.0.1:9080\\" --token=\\"\${{ github.sha }}\\" --team=\\"\${{ github.repository_owner }}\\"
    strategy: {}
"
`;
